sys/kern/vmem.c

```diff
@@ -270,12 +287,25 @@ int vmem_alloc(vmem_t *vm, vmem_size_t size, vmem_addr_t *addrp,
   WITH_MTX_LOCK (&vm->vm_lock) {
     vmem_check_sanity(vm);
 
+  retry:
     bt = bt_find_freeseg(vm, size);
 
     if (bt == NULL) {
-      pool_free(P_BT, btnew);
       klog("%s: block of %lu bytes not found in '%s'", __func__, size,
            vm->vm_name);
+      if (grow_called || !(vm->vm_flags & VM_GROW))
+        goto fail;
+
+      vaddr_t old = maxkvaddr;
+      maxkvaddr = pmap_growkernel(maxkvaddr + size);
+      klog("%s: increase kernel end %08lx -> %08lx", __func__, old, maxkvaddr);
+      grow_called = true;
+      if (vmem_add_nolock(vm, old, maxkvaddr - old) != 0)
+        goto fail;
+      goto retry;
```

Comment:
`vmem` is supposed to be a general-purpose allocator that can be used not only for kernel virtual address space allocation. This change tightly couples the `VM_GROW` flag with kernel virtual address space allocation, i.e. no one else can use the `VM_GROW` flag. I don't like this. Maybe the KVA-specific code could be moved to `kva_alloc()`?

Reply:
Ok, but we use `vmem` only as a backend for `kmem`. If I remember correctly @cahirwpz said that it is a correct place for calling `pmap_growkernel`.
I can move that code into `kva_alloc` but then from my point of view it will be more complicated.

Reply:
Hmmm... ok, I'm not going to argue with him.

Reply:
I will mark this conversation as unresolved and I will wait for his opinion - maybe I didn't understand him correctly.

Reply:
I have moved everything to `kmem`.